### ISSUE TITLE
Frontend issues and errors

### ISSUE THREAD

**Issue by aman-goel** [VeriGu/DistInv#1]
Hi,

I am trying to run DistAI on a set of Ivy protocols I compiled from different sources (https://github.com/aman-goel/ivybench/tree/temp). While I was able to successfully run DistAI on several of these protocols, DistAI failed on many of them, throwing different errors. I am suspecting most of these errors are due to minor issues with the DistAI frontend and are easy to resolve by the developers.

Here is a list of these protocols and their errors:

1. [quorum-leader-election](https://github.com/aman-goel/ivybench/blob/temp/ex/ivy_for_distai/quorum-leader-election.ivy), [hybrid_reliable_broadcast](https://github.com/aman-goel/ivybench/blob/temp/mypyv/ivy_for_distai/hybrid_reliable_broadcast.ivy)
````
Traceback (most recent call last):
  File "translate.py", line 1385, in <module>
    translate_ivy_to_python(PROBLEM)
  File "translate.py", line 1341, in translate_ivy_to_python
    parse_ivy_file('../protocols/{}/{}.ivy'.format(PROBLEM, PROBLEM))
  File "translate.py", line 872, in parse_ivy_file
    init_block[:0] = build_initialization_block()
  File "translate.py", line 255, in build_initialization_block
    lines.extend(default_axiom_rejection_sampling(default_axiom))
  File "translate.py", line 160, in default_axiom_rejection_sampling
    python_bexpr, python_stmts = ivy_bexpr_to_python_bexpr(axiom_str)
  File "translate.py", line 325, in ivy_bexpr_to_python_bexpr
    ivy_bexpr = remove_implication(ivy_bexpr)
  File "../DistAI/src-py/translate_helper.py", line 55, in remove_implication
    assert(len(dot_matches) <= 1)
AssertionError
````

2. [Simple](https://github.com/aman-goel/ivybench/blob/temp/tla/ivy_for_distai/Simple.ivy), [SimpleRegular](https://github.com/aman-goel/ivybench/blob/temp/tla/ivy_for_distai/SimpleRegular.ivy)
````
  File "Simple.py", line 123
    for , in permutations(value_indices):
        ^
SyntaxError: invalid syntax
````

3. [sharded_kv](https://github.com/aman-goel/ivybench/blob/temp/mypyv/ivy_for_distai/sharded_kv.ivy), [sharded_kv_no_lost_keys](https://github.com/aman-goel/ivybench/blob/temp/mypyv/ivy_for_distai/sharded_kv_no_lost_keys.ivy)
````
Ill-formed initialization statement: assume(owner(N1, K) & owner(N2, K) -> N1 = N2);
Can't open config file ../configs/sharded_kv_no_lost_keys.txt
````

4. [client_server_ae](https://github.com/aman-goel/ivybench/blob/temp/mypyv/ivy_for_distai/client_server_ae.ivy), [client_server_db_ae](https://github.com/aman-goel/ivybench/blob/temp/mypyv/ivy_for_distai/client_server_db_ae.ivy)
````
Traceback (most recent call last):
  File "translate.py", line 1385, in <module>
    translate_ivy_to_python(PROBLEM)
  File "translate.py", line 1342, in translate_ivy_to_python
    write_python_file('../auto_samplers/{}.py'.format(PROBLEM))
  File "translate.py", line 1292, in write_python_file
    python_codes.extend(build_sample_function())
  File "translate.py", line 1162, in build_sample_function
    enumerate_predicates()
  File "translate.py", line 1079, in enumerate_predicates
    var_base = type_abbrs[type_name]
KeyError: 'response'
````

5. [majorityset-leader-election](https://github.com/aman-goel/ivybench/blob/temp/ex/ivy_for_distai/majorityset-leader-election.ivy)
````
Unparsable line: trusted isolate ns = {
Can't open config file ../configs/majorityset-leader-election.txt
````

6. [learning_switch](https://github.com/aman-goel/ivybench/blob/temp/mypyv/ivy_for_distai/learning_switch.ivy)
````
Traceback (most recent call last):
  File "translate.py", line 1385, in <module>
    translate_ivy_to_python(PROBLEM)
  File "translate.py", line 1341, in translate_ivy_to_python
    parse_ivy_file('../protocols/{}/{}.ivy'.format(PROBLEM, PROBLEM))
  File "translate.py", line 879, in parse_ivy_file
    parse_action(action_str, action_buffer)
  File "translate.py", line 808, in parse_action
    python_stmts = parse_if_block(if_cond, if_stmts)
  File "translate.py", line 628, in parse_if_block
    python_bexpr, python_pre_stmts = ivy_bexpr_to_python_bexpr(if_cond)
  File "translate.py", line 331, in ivy_bexpr_to_python_bexpr
    pure_expr, uqvars, eqvars = extract_leading_quantifiers(sub_ivy_bexpr)
  File "../DistAI/src-py/translate_helper.py", line 109, in extract_leading_quantifiers
    assert(quantifier_part.startswith('exists '))
AssertionError
````

I hope this is of some help in future developments of DistAI. Please feel free to contact me for any details.

--
[Aman Goel](https://aman-goel.github.io/)

**Comment by jyao15** [comment]
Thanks for raising the issue and providing the examples. We will work on addressing these errors and get back to you as soon as possible.

**Comment by jyao15** [comment]
We have fixed several issues in the Ivy translator. Now DistAI is able to parse all 10 protocols except [majorityset-leader-election](https://github.com/aman-goel/ivybench/blob/temp/ex/ivy_for_distai/majorityset-leader-election.ivy), in which the "trusted isolate" appears difficult to handle in the current implementation. There seems to be a missing argument ```prev:node``` in action ```step_b``` of [SimpleRegular](https://github.com/aman-goel/ivybench/blob/temp/tla/ivy_for_distai/SimpleRegular.ivy). 

During this process we also realized another issue not directly shown by the examples above. When there was a "double declaration" such as in [distributed_lock](https://github.com/VeriGu/DistAI/blob/8c09127f0f0040a67a5b30dc6d1f33e84b977566/protocols/distributed_lock/distributed_lock.ivy) (```e``` is declared as an individual and an argument of action ```grant```), DistAI would have undefined behavior. Now DistAI will prompt the user to rename one of them.

Thanks again for helping to improve DistAI. Feel free to bring forward any further issue.
